Flaky integration test case "TestIPAssigner"

**Describe the bug**
<!--
A clear and concise description of what the bug is.
If you believe this bug is a security issue, please don't use this template and follow our [security guidelines](/SECURITY.md)
-->
A flaky integration test is observed with the below output,

```
=== RUN   TestIPAssigner
I1230 09:37:38.897460   13109 ip_assigner_linux.go:487] "Creating VLAN sub-interface" interface="antrea-ext.20" parent="eth0" vlan=20
I1230 09:37:38.898176   13109 ip_assigner_linux.go:100] "Assigned IP to interface" ip="10.10.20.11" interface="antrea-ext.20"
I1230 09:37:38.905726   13109 ip_assigner_linux.go:487] "Creating VLAN sub-interface" interface="antrea-ext.30" parent="eth0" vlan=30
I1230 09:37:38.906408   13109 ip_assigner_linux.go:100] "Assigned IP to interface" ip="10.10.30.10" interface="antrea-ext.30"
I1230 09:37:38.908090   13109 ip_assigner_linux.go:100] "Assigned IP to interface" ip="10.10.10.10" interface="antrea-dummy0"
I1230 09:37:38.925500   13109 ip_assigner_linux.go:100] "Assigned IP to interface" ip="10.10.10.11" interface="antrea-dummy0"
I1230 09:37:38.933412   13109 ip_assigner_linux.go:100] "Assigned IP to interface" ip="2021:124:6020:1006:250:56ff:fea7:36c2" interface="antrea-dummy0"
I1230 09:37:38.935295   13109 ip_assigner_linux.go:100] "Assigned IP to interface" ip="10.10.20.10" interface="antrea-ext.20"
I1230 09:37:38.965013   13109 ip_assigner_linux.go:100] "Assigned IP to interface" ip="2021:124:6020:1006:250:56ff:fea7:36c4" interface="antrea-dummy0"
I1230 09:37:38.965141   13109 ip_assigner_linux.go:145] "Deleted IP from interface" ip="2021:124:6020:1006:250:56ff:fea7:36c2" interface="antrea-dummy0"
I1230 09:37:38.965228   13109 ip_assigner_linux.go:145] "Deleted IP from interface" ip="10.10.20.11" interface="antrea-ext.20"
I1230 09:37:38.965306   13109 ip_assigner_linux.go:145] "Deleted IP from interface" ip="10.10.30.10" interface="antrea-ext.30"
I1230 09:37:38.965322   13109 ip_assigner_linux.go:406] "Deleting VLAN sub-interface" interface="antrea-ext.30" vlan=30
    ip_assigner_linux_test.go:101: 
        	Error Trace:	/usr/src/antrea.io/antrea/test/integration/agent/ip_assigner_linux_test.go:101
        	Error:      	Not equal: 
        	            	expected: sets.Set[string]{"10.10.20.10/24":sets.Empty{}}
        	            	actual  : sets.Set[string]{}
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1,4 +1,2 @@
        	            	-(sets.Set[string]) (len=1) {
        	            	- (string) (len=14) "10.10.20.10/24": (sets.Empty) {
        	            	- }
        	            	+(sets.Set[string]) {
        	            	 }
        	Test:       	TestIPAssigner
        	Messages:   	Actual IPs don't match
I1230 09:37:38.975390   13109 ip_assigner_linux.go:145] "Deleted IP from interface" ip="10.10.10.11" interface="antrea-dummy0"
I1230 09:37:38.975464   13109 ip_assigner_linux.go:145] "Deleted IP from interface" ip="2021:124:6020:1006:250:56ff:fea7:36c4" interface="antrea-dummy0"
I1230 09:37:38.975543   13109 ip_assigner_linux.go:142] "IP does not exist on interface" ip="10.10.20.10" interface="antrea-ext.20"
I1230 09:37:38.975557   13109 ip_assigner_linux.go:145] "Deleted IP from interface" ip="10.10.20.10" interface="antrea-ext.20"
I1230 09:37:38.975566   13109 ip_assigner_linux.go:406] "Deleting VLAN sub-interface" interface="antrea-ext.20" vlan=20
I1230 09:37:38.987030   13109 ip_assigner_linux.go:145] "Deleted IP from interface" ip="10.10.10.10" interface="antrea-dummy0"
--- FAIL: TestIPAssigner (0.13s)
```

links: https://github.com/antrea-io/antrea/actions/runs/12544675222/job/34977664966?pr=6889

 
**To Reproduce**
<!-- Steps to reproduce the behavior. -->

**Expected**
<!-- A clear and concise description of what you expected to happen. -->

**Actual behavior**
<!-- A clear and concise description of what's the actual behavior. If applicable, add screenshots, log messages, etc. to help explain the problem. -->

**Versions:**
<!--
Please provide the following information:
 - Antrea version (Docker image tag).
 - Kubernetes version (use `kubectl version`). If your Kubernetes components have [different versions](https://kubernetes.io/docs/setup/release/version-skew-policy/), please provide the version for all of them.
 - Container runtime: which runtime are you using (e.g. containerd, cri-o, docker) and which version are you using?
 - Linux kernel version on the Kubernetes Nodes (`uname -r`).
 - If you chose to compile the Open vSwitch kernel module manually instead of using the kernel module built into the Linux kernel, which version of the OVS kernel module are you using? Include the output of `modinfo openvswitch` for the Kubernetes Nodes.
-->

**Additional context**
<!-- Add any other context about the problem here, such as Antrea logs, kubelet logs, etc. -->

<!-- (Please consider pasting long output into a [GitHub gist](https://gist.github.com) or any other pastebin.) -->

--- a/pkg/agent/ipassigner/ip_assigner_linux.go
+++ b/pkg/agent/ipassigner/ip_assigner_linux.go
@@ -294,7 +294,6 @@ func getARPIgnoreForInterface(iface string) (int, error) {
 		return arpIgnore, nil
 	}
 	return arpIgnoreAll, nil
-
 }
 
 // ensureDummyDevice creates the dummy device if it doesn't exist.
@@ -306,7 +305,14 @@ func ensureDummyDevice(deviceName string) (netlink.Link, error) {
 	dummy := &netlink.Dummy{
 		LinkAttrs: netlink.LinkAttrs{Name: deviceName},
 	}
-	if err = netlink.LinkAdd(dummy); err != nil {
+	if err := netlink.LinkAdd(dummy); err != nil {
+		return nil, err
+	}
+	// When the primary IP address is removed from the interface, promote a corresponding secondary IP address
+	// instead of removing all the corresponding secondary IP addresses. Otherwise, the deletion of one IP address
+	// can trigger the automatic removal of all other IP addresses in the same subnet, if the deleted IP happens to
+	// be the primary (first one assigned chronologically).
+	if err := util.EnsurePromoteSecondariesOnInterface(deviceName); err != nil {
 		return nil, err
 	}
 	return dummy, nil
@@ -501,6 +507,13 @@ func (a *ipAssigner) getAssignee(subnetInfo *crdv1b1.SubnetInfo, createIfNotExis
 	// external interface when looking up the main table. To make it look up the custom table, we will need to restore
 	// the mark on the reply traffic and turn on src_valid_mark on this interface, which is more complicated.
 	if err := util.EnsureRPFilterOnInterface(name, 2); err != nil {
+		return nil, err
+	}
+	// When the primary IP address is removed from the interface, promote a corresponding secondary IP address
+	// instead of removing all the corresponding secondary IP addresses. Otherwise, the deletion of one IP address
+	// can trigger the automatic removal of all other IP addresses in the same subnet, if the deleted IP happens to
+	// be the primary (first one assigned chronologically).
+	if err := util.EnsurePromoteSecondariesOnInterface(name); err != nil {
 		return nil, err
 	}
 	as, err := a.addVLANAssignee(vlan, subnetInfo.VLAN)

--- a/pkg/agent/util/net_linux.go
+++ b/pkg/agent/util/net_linux.go
@@ -354,6 +354,11 @@ func EnsureRPFilterOnInterface(ifaceName string, value int) error {
 	return sysctl.EnsureSysctlNetValue(path, value)
 }
 
+func EnsurePromoteSecondariesOnInterface(ifaceName string) error {
+	path := fmt.Sprintf("ipv4/conf/%s/promote_secondaries", ifaceName)
+	return sysctl.EnsureSysctlNetValue(path, 1)
+}
+
 func getRoutesOnInterface(linkIndex int) ([]interface{}, error) {
 	link, err := netlinkUtil.LinkByIndex(linkIndex)
 	if err != nil {

--- a/test/integration/agent/ip_assigner_linux_test.go
+++ b/test/integration/agent/ip_assigner_linux_test.go
@@ -52,9 +52,43 @@ func TestIPAssigner(t *testing.T) {
 	ip1VLAN30 := "10.10.30.10"
 	subnet20 := &crdv1b1.SubnetInfo{PrefixLength: 24, VLAN: 20}
 	subnet30 := &crdv1b1.SubnetInfo{PrefixLength: 24, VLAN: 30}
-	desiredIPs := map[string]*crdv1b1.SubnetInfo{ip1: nil, ip2: nil, ip3: nil, ip1VLAN20: subnet20, ip2VLAN20: subnet20, ip1VLAN30: subnet30}
+	// These IPs will be assigned to the correct interface, in the specified order.
+	ipsToAssign := []struct {
+		ip         string
+		subnetInfo *crdv1b1.SubnetInfo
+	}{
+		{
+			ip: ip1,
+		},
+		{
+			ip: ip2,
+		},
+		{
+			ip: ip3,
+		},
+		// ip1VLAN20 and ip2VLAN20 are in the same subnet and will be assigned to the same
+		// interface (antrea-ext.20).
+		// ip1VLAN20 will be assigned first, which means ip1VLAN20 will be the "primary" IP,
+		// while ip2VLAN20 will be the "secondary" IP.
+		{
+			ip:         ip1VLAN20,
+			subnetInfo: subnet20,
+		},
+		{
+			ip:         ip2VLAN20,
+			subnetInfo: subnet20,
+		},
+		{
+			ip:         ip1VLAN30,
+			subnetInfo: subnet30,
+		},
+	}
 
-	for ip, subnetInfo := range desiredIPs {
+	desiredIPs := make(map[string]*crdv1b1.SubnetInfo)
+	for _, assignment := range ipsToAssign {
+		ip, subnetInfo := assignment.ip, assignment.subnetInfo
+		desiredIPs[ip] = subnetInfo
+
 		_, errAssign := ipAssigner.AssignIP(ip, subnetInfo, false)
 		cmd := exec.Command("ip", "addr")
 		out, err := cmd.CombinedOutput()
@@ -88,7 +122,14 @@ func TestIPAssigner(t *testing.T) {
 	assert.Equal(t, map[string]*crdv1b1.SubnetInfo{}, newIPAssigner.AssignedIPs(), "Assigned IPs don't match")
 
 	ip4 := "2021:124:6020:1006:250:56ff:fea7:36c4"
-	newDesiredIPs := map[string]*crdv1b1.SubnetInfo{ip1: nil, ip2: nil, ip4: nil, ip1VLAN20: subnet20}
+	// ip1VLAN20 is omitted, so it will be removed from the antrea-ext.20 interface. Because it
+	// is the primary IP address, secondary IPs (in this case ip2VLAN20) in the same subnet will
+	// be automatically removed when the primary is removed, unless the promote_secondaries
+	// sysctl variable has been set to 1 on the interface, which should be the case.
+	// By removing ip1VLAN20 (primary), we can therefore validate that IPAssigner is setting
+	// promote_secondaries correctly on the interface, as otherwise ip2VLAN20 will be removed
+	// automatically.
+	newDesiredIPs := map[string]*crdv1b1.SubnetInfo{ip1: nil, ip2: nil, ip4: nil, ip2VLAN20: subnet20}
 	err = newIPAssigner.InitIPs(newDesiredIPs)
 	require.NoError(t, err, "InitIPs failed")
 	assert.Equal(t, newDesiredIPs, newIPAssigner.AssignedIPs(), "Assigned IPs don't match")
@@ -98,7 +139,7 @@ func TestIPAssigner(t *testing.T) {
 	assert.Equal(t, sets.New[string](fmt.Sprintf("%s/32", ip1), fmt.Sprintf("%s/32", ip2), fmt.Sprintf("%s/128", ip4)), actualIPs, "Actual IPs don't match")
 	actualIPs, err = listIPAddresses(vlan20Device)
 	require.NoError(t, err, "Failed to list IP addresses")
-	assert.Equal(t, sets.New[string](fmt.Sprintf("%s/%d", ip1VLAN20, subnet20.PrefixLength)), actualIPs, "Actual IPs don't match")
+	assert.Equal(t, sets.New[string](fmt.Sprintf("%s/%d", ip2VLAN20, subnet20.PrefixLength)), actualIPs, "Actual IPs don't match")
 	_, err = netlink.LinkByName("antrea-ext.30")
 	require.Error(t, err, "VLAN 30 device should be deleted but was not")
 

